pkg/trace/agent/agent.go

```diff
@@ -484,11 +475,37 @@ func isManualUserDrop(priority sampler.SamplingPriority, pt *traceutil.Processed
 	return dm == manualSampling
 }
 
-// sample reports the number of events found in pt and whether the chunk should be kept as a trace.
-// sample does a semi-deep copy of pt to avoid making accidental changes to which spans are in the trace.
-// But any changes made directly to the spans, such as setting tags, etc. is not allowed.
-func (a *Agent) sample(now time.Time, ts *info.TagStats, pt *traceutil.ProcessedTrace) (numEvents int64, keep bool, retPt *traceutil.ProcessedTrace) {
-	pt = pt.Clone()
+// sample performs all sampling on the processedTrace modifying it as needed and returning if the trace should be kept and the number of events in the trace
+func (a *Agent) sample(now time.Time, ts *info.TagStats, pt *traceutil.ProcessedTrace) (keep bool, numEvents int) {
+	// We have a `keep` that is different from pt's `DroppedTrace` field as `DroppedTrace` will be sent to intake.
+	// For example: We want to maintain the overall trace level sampling decision for a trace with Analytics Events
+	// where a trace might be marked as DroppedTrace true, but we still sent analytics events in that ProcessedTrace.
+	keep, skipAnalyticsEvents := a.traceSampling(now, ts, pt)
+
+	var events []*pb.Span
+	if !skipAnalyticsEvents {
+		// We skip analytics events when a trace is marked as manual drop (aka priority -1)
+		// Note that we DON'T skip single span sampling. We only do this for historical
+		// reasons and analytics events are deprecated so hopefully this can all go away someday.
+		events = a.getAnalyzedEvents(pt, ts)
+	}
+	if !keep {
+		modified := sampler.SingleSpanSampling(pt)
+		if !modified {
+			// If there were no sampled spans, and we're not keeping the trace, let's use the analytics events
+			// This is OK because SSS is a replacement for analytics events so both should not be configured
+			// And when analytics events are fully gone we can get rid of all this
+			pt.TraceChunk.Spans = events
+		} else if len(events) > 0 {
+			log.Warnf("Detected both analytics events AND single span sampling in the same trace. Single span sampling wins because App Analytics is deprecated.")
+		}
```

Comment:
Is this behavior covered in a unit test?

Reply:
it is now!